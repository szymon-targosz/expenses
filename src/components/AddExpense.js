import React from 'react';

export default class AddExpense extends React.Component {
    render() {
        return (
            <div>
                AddExpense page!
            </div>
        );
    }
}