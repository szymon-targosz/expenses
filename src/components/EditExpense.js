import React from 'react';

export default class EditExpense extends React.Component {
    render() {
        return (
            <div>
                EditExpense page! Id number {this.props.match.params.id}.
            </div>
        );
    }
}