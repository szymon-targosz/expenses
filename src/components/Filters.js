import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class Filters extends React.Component {
    state = {
        calendarFocused: null
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        const val = e.target.value;
        if (val === 'date') {
          this.props.sortByDate();  
        } else if (val === 'amount') {
          this.props.sortByAmount();  
        };
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    render() {
        return (
            <div>
                <input 
                    type='text'
                    placeholder='Search'
                    value={this.props.filters.text}
                    onChange={this.onTextChange}    
                />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId='start'
                    endDateId='end'
                    endDate={this.props.filters.endDate}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    onDatesChange={this.onDatesChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStateToProps = state => ({
    filters: state.filters
});

const mapDispatchToProps = dispatch => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);