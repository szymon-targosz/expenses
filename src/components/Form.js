import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

export default class Form extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = e => {
        const amount = e.target.value;     
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        };
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        };
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf(),
                amount: parseFloat(this.state.amount, 10) * 100,
            });
        };
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input 
                    type="text" 
                    value={this.state.description} 
                    placeholder='Description' 
                    onChange={this.onDescriptionChange}
                    autoFocus
                />
                <input 
                    type="text" 
                    value={this.state.amount} 
                    placeholder='Amount' 
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea 
                    value={this.state.note}
                    placeholder='Add a note (optional)' 
                    onChange={this.onNoteChange}
                ></textarea>
                <button>Save Expense</button>
            </form>
        );
    }
}