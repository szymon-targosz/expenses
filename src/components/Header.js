import React from 'react';
import { Link } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';

export const Header = ({ startLogout }) => (
    <header>
        <Link to='/dashboard'>
            <h1>Expensify</h1>
        </Link>
        <button onClick={startLogout}>Logout</button>
    </header>
);

const mapDispatchToProps = dispatch => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);