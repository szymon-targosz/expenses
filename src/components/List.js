import React from 'react';
import ListItem from './ListItem';
import { connect } from 'react-redux';

export const List = (props) => (
    <div>
        {props.expenses.length === 0 ? (
            <div>no expenses</div>
        ) : (
            props.expenses.map(expense => <ListItem key={expense.id} {...expense} />)
        )}
    </div>
);

const mapStateToProps = (state) => ({
    expenses: state.expenses
});

export default connect(mapStateToProps)(List);