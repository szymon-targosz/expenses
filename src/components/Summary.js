import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expensesWithFilters';
import computeSum from '../selectors/expensesTotal';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

export const Summary = ({ expensesSum, expensesCount }) => {
    const matchingWord = expensesCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <h1>Viewing <span>{expensesCount}</span> {matchingWord} totalling <span>{numeral(expensesSum / 100).format('$0,0.00')}</span></h1>
            <Link className='button' to='/create'>Add Expense</Link>
        </div>
    );
};

const mapStateToProps = state => {
    const selectedExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesSum: computeSum(selectedExpenses),
        expensesCount: selectedExpenses.length 
    };
};

export default connect(mapStateToProps)(Summary);