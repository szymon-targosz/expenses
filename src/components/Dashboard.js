import React from 'react';
import List from './List';
import Filters from './Filters';
import Summary from './Summary';

const Dashboard = () => (
    <div>
        <Summary />
        <Filters />
        <List />
    </div>
);

export default Dashboard;