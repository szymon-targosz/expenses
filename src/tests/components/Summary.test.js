import React from 'react';
import { shallow } from 'enzyme';
import { Summary } from '../../components/Summary';

test('should correctly render Summary with multiple expenses', () => {
    const wrapper = shallow(<Summary expensesSum={1234} expensesCount={4}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render Summary with one expense', () => {
    const wrapper = shallow(<Summary expensesSum={100} expensesCount={1}/>);
    expect(wrapper).toMatchSnapshot();
});