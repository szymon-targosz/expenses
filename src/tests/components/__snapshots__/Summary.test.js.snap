// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Summary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    <span>
      4
    </span>
     
    expenses
     totalling 
    <span>
      1234
    </span>
  </h1>
</div>
`;

exports[`should correctly render Summary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    <span>
      1
    </span>
     
    expense
     totalling 
    <span>
      100
    </span>
  </h1>
</div>
`;
